src/core/graph.ts

```diff
@@ -410,10 +433,10 @@ export class DependencyGraph<T> {
   // have a graph without any leaf nodes.
   // As a result, starting a graph search only from leaf nodes in a
   // cyclic graph, will not necessarily traverse the entire graph.
-  public leafNodes(): Set<T> {
-    const leafs = new Set<T>(this.nodes());
-    for (const node of this.nodes()) {
-      for (const dep of this.getEdges(node)) {
+  public pureEffectNodes(): Set<T> {
```

Comment:
I'm not sure I like this change. While anyone who has dealt with graphs would know what a "leaf" or "root" node is, term "pure" is rather elusive.

Reply:
The issue with `leaf` or `root` terminology is that it is a bit confusing to know whether it is referring to leaf with respect to the underlying data structure or the direction of data flow.
I do think this naming is not as satisfactory and I'll think about that names could be unambiguous and descriptive.

Reply:
After a second thought, I think it would be a good idea to do so; as long as we consistently expose API names as a precedence graph I think it would be fine? We would need to document these API calls though, so that people are not confused.